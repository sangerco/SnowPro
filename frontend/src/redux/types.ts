export interface UserData {
    username: string;
    firstName: string;
    lastName: string;
    email: string;
    avatar: string;
    bio: string;
    videos: string[];
    photos: string[];
    favMountains: string[];
  }


export const FETCH_USER_DATA_REQUEST = 'FETCH_USER_DATA_REQUEST';
export const FETCH_USER_DATA_SUCCESS = 'FETCH_USER_DATA_SUCCESS';
export const FETCH_USER_DATA_FAILURE = 'FETCH_USER_DATA_FAILURE';


export interface FetchUserDataRequestAction {
    type: typeof FETCH_USER_DATA_REQUEST
}

export interface FetchUserDataSuccessAction {
    type: typeof FETCH_USER_DATA_SUCCESS;
    payload: UserData;
}

export interface FetchUserDataFailureAction {
    type: typeof FETCH_USER_DATA_FAILURE;
    payload: string;
}

export type UserActionTypes = 
    FetchUserDataRequestAction
    | FetchUserDataSuccessAction
    | FetchUserDataFailureAction;