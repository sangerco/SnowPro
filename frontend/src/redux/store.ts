import { Action, compose } from "redux";
import { ThunkAction, configureStore } from "@reduxjs/toolkit";
import { logger } from "redux-logger";
import rootReducer from "./reducers/rootReducer";

const composeEnhancers =
  (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;

export const store = configureStore({
  reducer: rootReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(logger, composeEnhancers),
});
