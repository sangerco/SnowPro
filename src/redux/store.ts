import { configureStore, combineReducers } from "@reduxjs/toolkit";
import userReducer from "./slices/userSlice";
import authReducer from "./slices/authSlice";
import messageReducer from "./slices/messageSlice";
import messageReplyReducer from "./slices/messageReplySlice";
import reviewReducer from "./slices/reviewSlice";
import reviewReplyReducer from "./slices/reviewReplySlice";
import tagReducer from "./slices/tagSlice";
import storage from "redux-persist/lib/storage";
import { persistReducer, persistStore } from "redux-persist";

const persistConfig = {
  key: "root",
  storage,
  blacklist: ["messages", "messageReplies", "reviews", "reviewReplies"],
};

const rootReducer = combineReducers({
  users: userReducer,
  auth: authReducer,
  messages: messageReducer,
  messageReplies: messageReplyReducer,
  reviews: reviewReducer,
  reviewReplies: reviewReplyReducer,
  tags: tagReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const store = configureStore({
  reducer: persistedReducer,
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const persistor = persistStore(store);
