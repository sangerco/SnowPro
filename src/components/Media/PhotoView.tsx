import React, { useEffect, useState } from "react";
import { deletePhoto, fetchPhoto } from "../../redux/slices/mediaSlices";
import { fetchAllTags } from "../../redux/slices/tagSlice";
import { useSelector, useDispatch } from "react-redux";
import { RootState, AppDispatch } from "../../redux/store";
import {
  Container,
  Image,
  Button,
  Icon,
  Divider,
  Dimmer,
  Loader,
  Header,
  Label,
  Modal,
} from "semantic-ui-react";
import { useNavigate } from "react-router";
import { Link } from "react-router-dom";

interface PhotoViewProps {
  id: string;
}

const PhotoView: React.FC<PhotoViewProps> = ({ id }) => {
  const dispatch = useDispatch<AppDispatch>();
  const navigate = useNavigate();
  const auth = useSelector((state: RootState) => state.auth);
  const userId = auth.data?.id;

  useEffect(() => {
    dispatch(fetchPhoto(id));
  }, [dispatch, id]);

  const media = useSelector((state: RootState) => state.media);
  const photo = media.photo;

  useEffect(() => {
    dispatch(fetchAllTags());
  }, [dispatch]);

  const tagData = useSelector((state: RootState) => state.tags);
  const tags = tagData.tags;

  let assocTags = [];

  if (tags && photo) {
    if (photo.tagIds && photo.tagIds.length > 0) {
      for (let i = 0; i > photo.tagIds.length; i++) {
        for (let j = 0; j > tags.length; j++) {
          if (photo.tagIds[i] === tags[j].id) {
            assocTags.push(tags[j].tag);
          }
        }
      }
    }
  }

  const [showDeleteModal, setShowDeleteModal] = useState(false);

  const handleDelete = (photoId: string) => {
    dispatch(deletePhoto(photoId));
  };

  if (tagData.loading || media.loading) {
    return (
      <Dimmer active>
        <Loader>Loading...</Loader>
      </Dimmer>
    );
  }

  if (media.error) {
    return (
      <Dimmer active>
        <Header as="h1">Error! Photo cannot be retrieved! {media.error}</Header>
        <Button color="red" onClick={() => navigate(-1)}>
          God Back
        </Button>
      </Dimmer>
    );
  }

  if (photo) {
    return (
      <div>
        <Container fluid>
          <Image src={photo.link} />
          <Divider />
          <Container text>
            <p>{photo.about}</p>
          </Container>
          {assocTags.map((tag) => (
            <Label color="green">{tag}</Label>
          ))}
          <Divider />
          <Button.Group>
            {userId === photo.userId ? (
              <>
                <Button onClick={() => setShowDeleteModal(true)}>
                  <Icon name="trash" />
                </Button>
                <Button as={Link} to={`'/media/photo/${photo.id}/update'`}>
                  <Icon name="edit" />
                </Button>
              </>
            ) : null}
          </Button.Group>
        </Container>
        {userId === photo.userId && (
          <Modal
            open={showDeleteModal}
            onClose={() => setShowDeleteModal(false)}>
            <Modal.Content>
              Are You Sure You Want To Delete Your Pic?
            </Modal.Content>
            <Modal.Actions>
              <Button negative onClick={() => handleDelete(photo.id)}>
                Yes
              </Button>
              <Button onClick={() => setShowDeleteModal(false)}>Cancel</Button>
            </Modal.Actions>
          </Modal>
        )}
      </div>
    );
  }

  return null;
};

export default PhotoView;
