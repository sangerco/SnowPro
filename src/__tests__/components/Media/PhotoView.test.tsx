/* eslint-disable testing-library/no-node-access */
import React from "react";
import PhotoView from "../../../components/Media/PhotoView";
import { render, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../../../redux/store";
import { BrowserRouter } from "react-router-dom";
import nock from "nock";

const propId = "fakePhotoId";

describe("test Photo View", () => {
  it("renders without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <PhotoView id={propId} />
        </BrowserRouter>
      </Provider>
    );
  });

  it("matches a snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <PhotoView id={propId} />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("renders a photo", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .get("/photo/photo-1")
      .reply(200, {
        photo: {
          id: "photo-1",
          userId: "testuser-1",
          username: "testuser",
          link: "https://www.testlinks.com/image1.jpg",
          about: "test image about",
          createdAt: "fake created date",
        },
      });
    render(
      <Provider store={store}>
        <BrowserRouter>
          <PhotoView id={"photo-1"} />
        </BrowserRouter>
      </Provider>
    );
    await waitFor(() => {
      const photo = screen.getByTestId("photo");
      expect(photo).toBeInTheDocument();
    });
    await waitFor(() => {
      const photoAbout = screen.getByTestId("photo-about");
      expect(photoAbout).toBeInTheDocument();
    });
  });
});
