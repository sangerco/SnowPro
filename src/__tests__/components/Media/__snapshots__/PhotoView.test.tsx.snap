// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Photo View matches a snapshot 1`] = `<DocumentFragment />`;
